File: languages/ruby/app_templates/kapp/appview.rb

=begin
  This is the main view class for %{APPNAME}.  Most of the non-menu,
  non-toolbar, and non-statusbar (e.g., non frame) GUI code should go
  here.
 
  This %{APPNAMELC} uses an HTML component as an example.
 
  @short Main view
  @author %{AUTHOR} <%{EMAIL}>
  @version %{VERSION}
=end
class %{APPNAME}View < Qt::Widget

    #
    # Use this signal to change the content of the statusbar
    #
    signals 'signalChangeStatusbar(const QString&)'

    #
    # Use this signal to change the content of the caption
    #
    signals 'signalChangeCaption(const QString&)'

    slots 'slotOnURL(const QString&)',
          'slotSetTitle(const QString&)'
   
    def initialize(parent)
        super(parent)
        
        # keep a reference to the DCOP Interface so it doesn't get gc'd
        @dcop = %{APPNAME}Iface.new(self)
        
        # setup our layout manager to automatically add our widgets
        top_layout = Qt::HBoxLayout.new(self)
        top_layout.setAutoAdd(true)
    
        # we want to look for all components that satisfy our needs.  the
        # trader will actually search through *all* registered KDE
        # applications and components -- not just KParts.  So we have to
        # specify two things: a service type and a constraint
        #
        # the service type is like a mime type.  we say that we want all
        # applications and components that can handle HTML -- 'text/html'
        #
        # however, by itself, this will return such things as Netscape..
        # not what we wanted.  so we constrain it by saying that the
        # string 'KParts/ReadOnlyPart' must be found in the ServiceTypes
        # field.  with this, only components of the type we want will be
        # returned.
        offers = KDE::Trader.self().query("text/html", "'KParts/ReadOnlyPart' in ServiceTypes")
    
        factory = nil
        # in theory, we only care about the first one.. but let's try all
        # offers just in when the first can't be loaded for some reason
        offers.each do |ptr|
            # we now know that our offer can handle HTML and is a part.
            # since it is a part, it must also have a library... let's try to
            # load that now
            factory = KDE::LibLoader.self().factory( ptr.library() )
            if ! factory.nil?
                @html = factory.create(self, ptr.name(), "KParts::ReadOnlyPart")
                break
            end
        end
    
        # if our factory is invalid, then we never found our component
        # and we might as well just exit now
        if factory.nil?
            KDE::MessageBox.error(self, i18n("Could not find a suitable HTML component"))
            return
        end
    
        connect(@html, SIGNAL('setWindowCaption(const QString&)'),
                self,   SLOT('slotSetTitle(const QString&)'))
        connect(@html, SIGNAL('setStatusBarText(const QString&)'),
                self,   SLOT('slotOnURL(const QString&)'))
    
    end
    
    
    def print(p, height, width)
        # do the actual printing, here
        # p.drawText(etc..)
    end
    
    def currentURL()
        return @html.url().url()
    end
    
    def openURL(url)
        @html.openURL(KDE::URL.new(url))
    end
    
    def slotOnURL(url)
        emit signalChangeStatusbar(url)
    end
    
    def slotSetTitle(title)
        emit signalChangeCaption(title)
    end

end

File: languages/cpp/app_templates/kapp/appview.cpp
%{CPP_TEMPLATE}

#include "%{APPNAMELC}view.h"

#include <qpainter.h>
#include <qlayout.h>

#include <kurl.h>

#include <ktrader.h>
#include <klibloader.h>
#include <kmessagebox.h>
#include <krun.h>
#include <klocale.h>

%{APPNAME}View::%{APPNAME}View(QWidget *parent)
    : QWidget(parent),
      DCOPObject("%{APPNAME}Iface")
{
    // setup our layout manager to automatically add our widgets
    QHBoxLayout *top_layout = new QHBoxLayout(this);
    top_layout->setAutoAdd(true);

    // we want to look for all components that satisfy our needs.  the
    // trader will actually search through *all* registered KDE
    // applications and components -- not just KParts.  So we have to
    // specify two things: a service type and a constraint
    //
    // the service type is like a mime type.  we say that we want all
    // applications and components that can handle HTML -- 'text/html'
    //
    // however, by itself, this will return such things as Netscape..
    // not what we wanted.  so we constrain it by saying that the
    // string 'KParts/ReadOnlyPart' must be found in the ServiceTypes
    // field.  with this, only components of the type we want will be
    // returned.
    KTrader::OfferList offers = KTrader::self()->query("text/html", "'KParts/ReadOnlyPart' in ServiceTypes");

    KLibFactory *factory = 0;
    // in theory, we only care about the first one.. but let's try all
    // offers just in case the first can't be loaded for some reason
    KTrader::OfferList::Iterator it(offers.begin());
    for( ; it != offers.end(); ++it)
    {
        KService::Ptr ptr = (*it);

        // we now know that our offer can handle HTML and is a part.
        // since it is a part, it must also have a library... let's try to
        // load that now
        factory = KLibLoader::self()->factory( ptr->library() );
        if (factory)
        {
            m_html = static_cast<KParts::ReadOnlyPart *>(factory->create(this, ptr->name(), "KParts::ReadOnlyPart"));
            break;
        }
    }

    // if our factory is invalid, then we never found our component
    // and we might as well just exit now
    if (!factory)
    {
        KMessageBox::error(this, i18n("Could not find a suitable HTML component"));
        return;
    }

    connect(m_html, SIGNAL(setWindowCaption(const QString&)),
            this,   SLOT(slotSetTitle(const QString&)));
    connect(m_html, SIGNAL(setStatusBarText(const QString&)),
            this,   SLOT(slotOnURL(const QString&)));

}

%{APPNAME}View::~%{APPNAME}View()
{
}

void %{APPNAME}View::print(QPainter *p, int height, int width)
{
    // do the actual printing, here
    // p->drawText(etc..)
}

QString %{APPNAME}View::currentURL()
{
    return m_html->url().url();
}

void %{APPNAME}View::openURL(QString url)
{
    openURL(KURL(url));
}

void %{APPNAME}View::openURL(const KURL& url)
{
    m_html->openURL(url);
}

void %{APPNAME}View::slotOnURL(const QString& url)
{
    emit signalChangeStatusbar(url);
}

void %{APPNAME}View::slotSetTitle(const QString& title)
{
    emit signalChangeCaption(title);
}
#include "%{APPNAMELC}view.moc"
